selinux: allow Cockpit to use HTTP keytab on IPA servers

abbra opened the pull request:
Cockpit can use GSSAPI authentication and has pretty good definition of how to enable it: https://cockpit-project.org/guide/latest/sso.html. These instructions work on IPA clients but they cannot be used on IPA servers because IPA framework already owns HTTP/.. Kerberos service and its keytab.

Luckily, there are two changes that need to be done to enable Cockpit single sign-on with GSSAPI on IPA servers:

  - create a symlink /etc/cockpit/krb5.keytab to /var/lib/ipa/gssproxy/http.keytab

  - add SELinux policy to allow cockpit_session_t to operate on ipa_var_lib_t files

No additional changes are needed. Note that Cockpit instructions above also talk about Kerberos service modifications to enable delegation. These modifications should not be done for IPA servers' HTTP services, as these services are already enabled to handle delegation.

Fixes: https://pagure.io/freeipa/issue/9675

rcritten commented:
This will give it access to much more than only the keytab. 

find /var/lib/ipa -context unconfined_u:object_r:ipa_var_lib_t:s0

abbra commented:
Correct. Unfortunately, there is no easier fix for this other than explicitly marking the keytab and making that context included in `ipa_var_lib_t` access as well. This will break existing installations as they'll need re-labelling.

On cockpit side, they have explicit location for the keytab (`/etc/cockpit/krb5.keytab`) so it is really about getting access to resolve a symlink properly. They never touch anything else.

abbra commented:
Additionally, `cockpit-ws` will run as root itself so it is already capable to access our paths, if needed.

rcritten commented:
Among what it would be allowed to read, for example, is the Apache private key PIN and the DNSSEC keys. So this isn't about what Cockpit runs as but what is potentially visible to it. If a bug in Cockpit allowed other files to be read then this leads to disclosure of secrets.

abbra commented:
I separated `http.keytab` into a different context and added two policy changes:
- updated `ipa_read_lib()` to grant reading `ipa_http_keytab_t`
- created `ipa_read_http_keytab()` to allow reading `ipa_http_keytab_t`. The latter requires directory access to `ipa_var_lib_t` directories, so it is included as well

On top of that, we already had code to migrate HTTP keytab during upgrade and this code is always called and chowns `http.keytab` file to the right user. I extended it to restore SELinux context to new value as well.

abbra commented:
Successful test run: http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/f61481be-8561-11ef-92aa-fa163e863465/report.html?sort=result
```
INFO     ipatests.pytest_ipa.integration.host.Host.master.IPAOpenSSHTransport:transport.py:391 RUN ['ipa-getcert', 'request', '-f', '/etc/cockpit/ws-certs.d/99-cockpit.cert', '-k', '/etc/cockpit/ws-certs.d/99-cockpit.key', '-D', 'master.ipa.test', '-K', 'host/master.ipa.test', '-m', '0640', '-o', 'root:cockpit-ws', '-O', 'root:root', '-M', '0644']
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd63:transport.py:513 RUN ['ipa-getcert', 'request', '-f', '/etc/cockpit/ws-certs.d/99-cockpit.cert', '-k', '/etc/cockpit/ws-certs.d/99-cockpit.key', '-D', 'master.ipa.test', '-K', 'host/master.ipa.test', '-m', '0640', '-o', 'root:cockpit-ws', '-O', 'root:root', '-M', '0644']
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd63:transport.py:557 New signing request "20241008105119" added.
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd63:transport.py:217 Exit code: 0
INFO     ipatests.pytest_ipa.integration.host.Host.master.IPAOpenSSHTransport:transport.py:391 RUN ['ln', '-s', '/var/lib/ipa/gssproxy/http.keytab', '/etc/cockpit/krb5.keytab']
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd64:transport.py:513 RUN ['ln', '-s', '/var/lib/ipa/gssproxy/http.keytab', '/etc/cockpit/krb5.keytab']
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd64:transport.py:217 Exit code: 0
INFO     ipatests.pytest_ipa.integration.host.Host.master.IPAOpenSSHTransport:transport.py:391 RUN ['systemctl', 'enable', '--now', 'cockpit.socket']
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd65:transport.py:513 RUN ['systemctl', 'enable', '--now', 'cockpit.socket']
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd65:transport.py:557 Created symlink /etc/systemd/system/sockets.target.wants/cockpit.socket → /usr/lib/systemd/system/cockpit.socket.
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd65:transport.py:217 Exit code: 0

------------------------------ Captured log call -------------------------------
INFO     ipatests.pytest_ipa.integration.host.Host.master.IPAOpenSSHTransport:transport.py:391 RUN ['kinit', 'admin']
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd66:transport.py:513 RUN ['kinit', 'admin']
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd66:transport.py:557 Password for admin@IPA.TEST: 
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd66:transport.py:217 Exit code: 0
INFO     ipatests.pytest_ipa.integration.host.Host.master.IPAOpenSSHTransport:transport.py:391 RUN ['/usr/bin/curl', '-u:', '--negotiate', '--cacert', '/etc/ipa/ca.crt', 'https://master.ipa.test:9090/cockpit/login']
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd67:transport.py:513 RUN ['/usr/bin/curl', '-u:', '--negotiate', '--cacert', '/etc/ipa/ca.crt', 'https://master.ipa.test:9090/cockpit/login']
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd67:transport.py:557   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd67:transport.py:557                                  Dload  Upload   Total   Spent    Left  Speed
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd67:transport.py:557 
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
100    81  100    81    0     0     61      0  0:00:01  0:00:01 --:--:--    61
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd67:transport.py:557 {"csrf-token":"b1ce17bd2f554f3b2df2172558278c38eb5258c4d80d16a8759fa8bad127c943"}
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd67:transport.py:217 Exit code: 0
```

rcritten commented on ipatests/test_integration/test_cockpit.py:
```diff
@@ -0,0 +1,61 @@
+#
+# Copyright (C) 2016  FreeIPA Contributors see COPYING for license
```
nit: date should be 2024.

rjeffman replied:
2024 seems to be enough for US, 2016-2024 may be preferred in other countries.

rcritten commented:
Adding ack since this change looks good. Waiting for CI to finish.

flo-renaud commented:
master:

* c775de3c2bf05b447bfd17646306f62406ffc6dc selinux: allow Cockpit to use HTTP keytab on IPA servers
* 4519c2fde183d8b8c4f49da37fed68a41a220d72 Minimal test for Cockpit integration on IPA master
* c4f3d9034ddfcddcb13e75d1c149d38da34dea08 Extend nightly tests with Cockpit test
